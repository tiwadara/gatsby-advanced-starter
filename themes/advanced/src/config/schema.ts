export const schema = `#graphql
  type WebsiteData {
    title: String!
    titleShort: String!
    name: String!
    description: String!
    language: String!
    logoUrl: String!
    fbAppId: String
    twitterName: String
    url: String!
    copyright: String!
    rss: String!
    rssTitle: String!
    googleAnalyticsId: String
    disqusShortname: String
    themeColor: String!
    backgroundColor: String!
  }

  type UserData {
    id: String!
    firstName: String!
    lastName: String!
    twitterName: String
    linkedIn: String
    github: String
    email: String!
    location: String!
    about: String!
    avatar: String!
  }

  type OrganizationData {
    name: String!
    description: String!
    logoUrl: String!
    url: String!
  }

  type IconManifest {
    src: String!
    sizes: String!
    type: String!
    purpose: String
  }


  type  SiteConfig {
    website: WebsiteData!
    user: UserData
    organization: OrganizationData

    pathPrefix: String!

    contentDir: String
    assetDir: String

    embeddedImageWidth: Int!
    embeddedVideoWidth: Int!

    iconPath: String
    iconList: [IconManifest]!
    iconCachePaths: [String]

    basePath: String
  }

  type SiteSiteMetadata {
    config: SiteConfig!
  }
`;

export default schema;
