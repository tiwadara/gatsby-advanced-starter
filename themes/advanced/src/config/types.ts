export interface WebsiteData {
  title: string;
  titleShort: string;
  name: string;
  description: string;
  language: string;
  logoUrl: string;
  fbAppId?: string;
  twitterName?: string;
  url: string;
  copyright: string;
  rss: string;
  rssTitle: string;
  googleAnalyticsId?: string;
  disqusShortname?: string;
  themeColor: string;
  backgroundColor: string;
}

export interface UserData {
  id: string;
  firstName: string;
  lastName: string;
  twitterName?: string;
  linkedIn?: string;
  github?: string;
  email: string;
  location: string;
  about: string;
  avatar: string;
}

export interface OrganizationData {
  name: string;
  description: string;
  logoUrl: string;
  url: string;
}

type IconManifest = {
  src: string;
  sizes: string;
  type: string;
  purpose?: string;
};

export interface SiteConfig {
  website: WebsiteData;
  user?: UserData;
  organization?: OrganizationData;

  pathPrefix: string;

  contentDir?: string;
  assetDir?: string;

  embeddedImageWidth: number;
  embeddedVideoWidth: number;

  iconPath?: string;
  iconList: Readonly<Array<IconManifest>>;
  iconCachePaths?: string[];

  basePath?: string;
}
