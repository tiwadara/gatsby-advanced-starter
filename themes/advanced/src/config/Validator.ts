import isAbsoluteUrl from "is-absolute-url";
import { SiteConfig } from "./types";

// Validate
const validateSiteConfig = (config: SiteConfig): Readonly<SiteConfig> => {
  const newConfig = config;
  // Make sure pathPrefix is empty if not needed
  if (newConfig.pathPrefix === "/") {
    newConfig.pathPrefix = "";
  } else {
    // Make sure pathPrefix only contains the first forward slash
    newConfig.pathPrefix = `/${newConfig.pathPrefix.replace(/^\/|\/$/g, "")}`;
  }

  // Make sure website.url doesn't have an ending forward slash
  if (newConfig.website.url.substr(-1) === "/")
    newConfig.website.url = newConfig.website.url.slice(0, -1);

  // Make sure website.rss has a starting forward slash
  if (newConfig.website.rss && newConfig.website.rss[0] !== "/")
    newConfig.website.rss = `/${newConfig.website.rss}`;

  // Make sure that website.url is an absolute URL
  if (!isAbsoluteUrl(newConfig.website.url)) {
    throw new Error("SiteConfig.website.url is not absolute.");
  }

  // Make sure that organization.url is an absolute URL
  if (newConfig.organization && !isAbsoluteUrl(newConfig.organization.url)) {
    throw new Error("SiteConfig.organization.url is not absolute.");
  }

  return newConfig;
};

export default validateSiteConfig;
