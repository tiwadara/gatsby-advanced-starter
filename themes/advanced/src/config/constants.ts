const constants = {
  postsPerFeedPage: 5, // Posts per feed page
  feedMetaDirectory: "feed_meta/", // Directory for storing the feed page metadata
};

export default constants;
