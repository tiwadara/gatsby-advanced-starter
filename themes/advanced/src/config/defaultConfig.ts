import { SiteConfig } from "./types";

const config: SiteConfig = {
  // Website configuration
  website: {
    title: "Gatsby Advanced Starter", // Homepage title
    titleShort: "Advanced Blog", // Short site title for homescreen (PWA). Preferably should be under 12 characters to prevent truncation
    name: "Gatsby Advanced Starter", // Website name used for homescreen (PWA) and SEO
    description: "A GatsbyJS starter equipped with advanced features.", // Website description used for RSS feeds/meta description tag
    language: "en", // Sets the global HTML lang attribute
    logoUrl: "/logos/logo-1024.png", // Logo used for SEO
    url: "https://example.com", // Domain of your website without the pathPrefix
    rss: "/rss.xml", // Path to the RSS file
    rssTitle: "Gatsby Advanced Starter RSS Feed", // Title of the RSS feed
    copyright: "Â© Copyright 2021", // Copyright string for the footer of the website and RSS feed.
    themeColor: "#D83850", // Used for setting manifest and progress theme colors.
    backgroundColor: "#F7F7F7", // Used for setting manifest background color.
  },

  // User configuration
  user: undefined,

  // Organization information used for SEO
  organization: undefined,

  // Gatsby Configuration
  pathPrefix: "/", // Prefixes all links. For cases when deployed to example.github.io/gatsby-advanced-starter/.

  contentDir: undefined, // Directory for MDX posts
  assetDir: undefined, // Asset directory

  embeddedImageWidth: 768, // MDX embedded image width. Used by gatsby-plugin-image for optimization
  embeddedVideoWidth: 920, // MDX embedded video width in pixels

  iconPath: undefined, // Icon used for manifest icon creation.
  iconList: [], // Icons used for the web manifest. Can be used to override iconPath for a more pixel perfect control.
  iconCachePaths: undefined, // Glob pattern path for the icons to be cached by the gatsby-plugin-offline

  basePath: undefined, // Base path for mounting pages. Allows for multiple themes to be used in a single website.
};

export default config;
