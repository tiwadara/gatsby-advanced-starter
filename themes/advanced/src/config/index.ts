import urlJoin from "url-join";
import { merge } from "lodash";

import validateSiteConfig from "./Validator";
import baseConfig from "./defaultConfig";
import { SiteConfig } from "./types";

export type {
  SiteConfig,
  OrganizationData,
  UserData,
  WebsiteData,
} from "./types";

export { default as schema } from "./schema";
export { default as constants } from "./constants";

export const withDefaults = (userConfig: SiteConfig): Readonly<SiteConfig> =>
  validateSiteConfig(merge(baseConfig, userConfig));

export const withBasePath = (config: SiteConfig, url: string): string =>
  config.basePath ? urlJoin(config.basePath, url) : url;

export { default as useConfig } from "./useConfig";
