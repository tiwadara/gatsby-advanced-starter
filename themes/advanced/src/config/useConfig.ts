import { useStaticQuery, graphql } from "gatsby";
import { SiteConfig } from "./types";

type UserConfigQueryType = {
  site?: {
    siteMetadata?: {
      config?: SiteConfig;
    };
  };
};

const useConfig = (): SiteConfig => {
  const resp = useStaticQuery<UserConfigQueryType>(
    graphql`
      query UserConfig {
        site {
          siteMetadata {
            config {
              contentDir
              assetDir
              embeddedImageWidth
              embeddedVideoWidth
              basePath
              iconPath
              iconCachePaths
              iconList {
                src
                sizes
                type
                purpose
              }
              organization {
                description
                logoUrl
                name
                url
              }
              pathPrefix
              user {
                about
                avatar
                firstName
                github
                email
                id
                lastName
                linkedIn
                location
                twitterName
              }
              website {
                backgroundColor
                copyright
                description
                language
                fbAppId
                googleAnalyticsId
                disqusShortname
                logoUrl
                name
                rss
                rssTitle
                themeColor
                title
                titleShort
                twitterName
                url
              }
            }
          }
        }
      }
    `
  );

  const config = resp.site?.siteMetadata?.config;

  if (!config) throw Error("useConfig: Failed to query SiteConfig.");

  return config;
};

export default useConfig;
