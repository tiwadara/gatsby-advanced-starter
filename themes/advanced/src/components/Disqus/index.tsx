import React from "react";
import { Disqus as DisqusPlugin } from "gatsby-plugin-disqus";

import { useConfig } from "../../config";
import { Post } from "../../types";

type DisqusProps = {
  post: Post;
};

const Disqus = ({ post }: DisqusProps): JSX.Element | null => {
  const config = useConfig();

  // Do not render if no shortname was provided
  if (!config.website.disqusShortname) {
    return null;
  }

  const { url, title } = post;

  return (
    <DisqusPlugin
      config={{
        url,
        identifier: title,
        title,
      }}
    />
  );
};

export default Disqus;
