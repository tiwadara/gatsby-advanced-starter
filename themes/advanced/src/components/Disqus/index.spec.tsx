import React from "react";
import { render } from "@testing-library/react";
import cloneDeep from "clone-deep";
import { mocked } from "ts-jest/utils";

import Disqus from "./index";
import { useConfig } from "../../config";

import { post, config as configFixture } from "../../../../test/fixtures";

jest.mock("../../config/useConfig");

const mockedUseConfig = mocked(useConfig);

describe("component Disqus", () => {
  it("renders correctly when config.disqusShortname is set", () => {
    expect.assertions(1);

    const newConfig = cloneDeep(configFixture);
    newConfig.website.disqusShortname =
      "https-vagr9k-github-io-gatsby-advanced-starter";

    mockedUseConfig.mockReturnValue(newConfig);

    const { asFragment } = render(<Disqus post={post} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly when config.disqusShortname is not set", () => {
    expect.assertions(1);

    const newConfig = cloneDeep(configFixture);
    newConfig.website.disqusShortname = undefined;

    mockedUseConfig.mockReturnValue(newConfig);

    const { asFragment } = render(<Disqus post={post} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
