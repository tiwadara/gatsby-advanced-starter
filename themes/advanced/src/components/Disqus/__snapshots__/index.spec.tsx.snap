// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Disqus renders correctly when config.disqusShortname is not set 1`] = `<DocumentFragment />`;

exports[`component Disqus renders correctly when config.disqusShortname is set 1`] = `
<DocumentFragment>
  <div
    id="disqus_thread"
  />
</DocumentFragment>
`;
